### DISCUSSION TITLE
Suitability for large datasets

### DISCUSSION THREAD

**Discussion by roaldarbol** [TheoMichelot/hmmTMB#22] [Q&A] [answered]
Hi there! I've been following along on the sidelines for a while now, and been on the fence about whether or not I needed to take the plunge into HMMs - now I think time is ripe. :-)

I'm working on some really massive data sets, videos recorded at 30fps for 24h, so ~2500000 data points for each individual animal, that have been tracked to extract the coordinates. I'm then hoping to find a good (semi)automated method for finding passive/active periods based on the velocity. Would this package be suitable for this? This is also a question about the compute time, have you had experience with data sets in this magnitude?

Looking forward to hearing back!

**Comment by TheoMichelot** [comment, marked as answer]
Hi, here are a few thoughts about using hmmTMB on large data sets.

1. We wrote the package with speed in mind, e.g., the likelihood function is implemented in C++ using TMB. That being said, we also wanted hmmTMB to be flexible (e.g., many observation distributions, general covariate dependence), so we had to sacrifice speed in some places.

2. I just ran a simulation to fit a 2-state Gaussian HMM to 2.5 million data points, and it took around 3 minutes on my laptop. Here I was fitting the true model, so it is likely that it would take somewhat longer in a real data analysis, but it should be feasible.

3. Computing time will very strongly depend on the model formulation, and model fitting can be much much slower for more complex models. In particular, it will depend on the following.
     - The number of states: With more states, model fitting is slower and more parameters need to be estimated. For comparison, I just fitted a 3-state Gaussian HMM to 1 million observations (40% of the previous simulation), and it took around 3.5 min.
     - Covariates: When covariates are included, more parameters need to be estimated, and things are slower.
     - Random effects: If random effects and/or non-linear effects are included in any part of the model, TMB has to integrate those out of the likelihood, and everything is much slower.
     - Observation distribution: Common distributions (e.g., normal, gamma, exponential, Poisson) use fast TMB implementations, so I would expect them to be faster than less standard distributions, which we had to implement from scratch (e.g., von Mises).
   
4. I'm not exactly sure what the reason is, but it seems that memory is sometimes a bottleneck for large data sets and complex models. For example, I tried running a 3-state HMM to 2.5 million observations, and my R session crashed after I ran out of memory (16Gb). I think this happens when TMB builds the computational graph to set up automatic differentiation, and the memory is only used briefly before model fitting starts, but in some cases it might be limiting.

5. You can fit the model to a subset of the data, and then use the fitted model to get the most likely state sequence for the rest of the data. This is described in Section 2 of the ["Advanced features of hmmTMB" vignette](https://github.com/TheoMichelot/hmmTMB/blob/master/vignettes/hmmTMB_advanced_features.pdf). This might be the best option in many large-data situations, if the main objective is to get the decoded state sequence.

Anecdotally, my impression is that very high-frequency data like yours are often pre-processed before applying a hidden Markov model. This is both to make things faster, and to extract the variables that we think will be the most informative to identify the states that we are interested in (e.g., mean or variance over 30-sec windows might be more useful than raw 30-Hz observations). But of course, this depends a lot on the specific application.

I hope this helps, I would be interested to hear how your analysis goes!

**Reply by roaldarbol** [reply]
Brilliant, thanks a bunch for the thoughts and suggestions! 

I also got started with running it and I manage in reasonable times too - am positively surprised! For now, I'm happy enough to continue with just 2 states to keep things snappy. I am working with many different individuals; would you suggest fitting ID as a random factor? And can you point me to some documentation regarding how to make use of formulas in general?

I'm trying to wrap my head around how to best specify `par`. Is there an *a priori* way of specifying them initially? Or do you take means/SD from distributions? I'm currently encountering two things, one I think you might be able to help with - the other is plain weird. Both are related to `par`:

1. I'm getting reasonable predictions(!), but I get quite a few false positives where the trace is clearly not active, but a single frame gets identified as being active. Are there ways I can tweak that?
2. I'm getting a super strange error! When trying to use `obs$suggest_initial()` I get: `Error in obs$suggest_initial() : object 'foo' not found`. I tried running the example from the docs, and that works just fine... searched your code and couldn't find `foo` anywhere, so I'm stumped...

As for your last point, that is actually one of the things I'm quite interested in looking at, i.e. the hierarchical embedding of behaviour. It's picking up on old ideas by Tinbergen, Dawkins and others, much prior to the computing power we have available today. I'm working at the intersection of quantitative ethology and neuroscience, so I'm trying to identify states at different levels/ across timescales without specifying particular behaviours. Happy to chat more about it at later time - could probably be interesting to collaborate on if it could pique your interest. :-)

**Reply by TheoMichelot** [reply]
Including a random intercept for ID would be helpful to allow for inter-individual heterogeneity, but it will make model fitting much slower. (The technical details are in [Section 2.2.1 of this paper](https://arxiv.org/pdf/2211.14139.pdf).) If you fit separate models to the different individuals, this isn't necessary anyway.

Here are some links for documentation on formulas:
- Section 7 of [the main hmmTMB vignette](https://github.com/TheoMichelot/hmmTMB/blob/master/vignettes/hmmTMB_workflow.pdf) describes the inclusion of covariates with a few examples, including random effects.
- I wrote [a short vignette describing the formula syntax for common models](https://github.com/TheoMichelot/hmmTMB/blob/master/vignettes/hmmTMB_formula_syntax.pdf), including random effects. The syntax is based on mgcv, so the documentation for that package is helpful too.

The choice of initial parameters (for `par`) should be based on the data and on what you expect the states to capture. I wrote a little bit about this in Section 5 of the '[Advanced features of hmmTMB](https://github.com/TheoMichelot/hmmTMB/blob/master/vignettes/hmmTMB_advanced_features.pdf)' vignette. Typically, you would want to take an educated guess at what the parameters might be, so that the numerical optimiser starts reasonably close to the parameter estimates.

Regarding your two specific questions:
1. I'm not sure I understand. If the Viterbi sequence switches to another state just for one time step, this seems to indicate that: (a) the observation at that time step suggests a switch, or (b) the transition probabilities make a switch seem likely at that time step, or (c) there is some numerical problem.
2. Yes, you are right, there used to be a bug in `Observation$suggest_initial()`. I fixed it a few months ago (commit 85abab9). You need to install the updated package: `devtools::install_github("TheoMichelot/hmmTMB")`

Also happy to keep in touch about this!

**Reply by roaldarbol** [reply]
That's super helpful, thanks! I think I'll just fit for each animal separately then, seems like less hassle. I know you've worked quite a bit with animal paths, have you tried using just velocities before? I'm trying to figure out whether normal distributions will work alright - it's of course quite skewed toward 0-ish values as there's a lot of rest in there, so I'm thinking that a gamma distribution might be a better fit. What are your thoughts on this?

Re. the Viterbi predictions it might just be some parameter tweaking that's needed, I don't think there are any numerical problems, it's likely that I just need to get better acquainted with the ins and outs of HMMs. For now, I'll just go with the outputs as they overall seem reasonable, and much better than what I was getting with simpler algorithms, thresholds, etc.

Also a few notes:
- `hmmTMB:::dist_list` which is suggested in the docs doesn't exist (even though I've updated to the Github version)
- I'm getting another error in `Observation$suggest_initial()`, but I'll make an issue for that separately.

Oh, and as a last Q, have you come across HMMs being used on data sets of this size before? Would be nice to have as a point in my PhD thesis if we're pushing some boundaries. :-) I'm really excited to have gotten started with `hmmTMB` and HMMs in general, I think they'll be a tool I'll make a lot of use of! Have you given an introductory talk or similar that can be watched on Youtube? Would love to hear that!

**Reply by TheoMichelot** [reply]
By velocity, do you mean a 2d (or 3d) vector, or do you mean the (1d) speed? If it's the latter, then I agree that something like a gamma distribution would be ideal. The `gamma2` distribution in hmmTMB is specified in terms of mean and standard deviation, so it's easy to switch from the normal.

I can't reproduce the error for `hmmTMB:::dist_list`. Any chance you accidentally wrote `hmmTMB::dist_list` (with only two colons)? If not, I'll need to look into it or at least fix the documentation.

I can't think of HMM studies with millions of data points off the top of my head, but that doesn't mean they're not out there! Within the field of applications of HMMs to animal tracking, that's certainly a very large data set.

I've given a couple of talks on HMMs that are available online:
- "[Movement ecology: introduction to analysing animal movement data in R](https://www.youtube.com/watch?v=WELTpbB5BuU)" was a webinar for the Ecological Forecasting Initiative. I talk about standard HMMs for x-y animal movement data and common challenges, and go through a tutorial for the package momentuHMM (another general HMM package)
- "[Flexible hidden Markov models using hmmTMB and applications in ecology](https://cassyni.com/events/71QFHpfprvkFTkRGCdvLQu)" is about hmmTMB, with a few examples for tracking data, human acceleration data, and occupancy data.

**Reply by roaldarbol** [reply]
Yes, it's 1D. The animals (beetles) were in small circular dishes in the lab, so there's nothing really meaningful about the directionality (and at the time of tracking, I couldn't get the orientation of the animal - for future work, I'll be able to have that too, so that might change the levels I'll use). Great suggestion with the `gamma2` distribution, I'll give that a go!

Sorry, that was my bad completely - with `:::` it works. 

I'll have a look around, but even just for animal movement data, that's already good to know - thanks! And also for the links, that'll be my dinner-TV tonight.😉 Thanks for taking the time to respond so quickly and thoroughly, I really appreciate it!

**Reply by roaldarbol** [reply]
Just an update on my end - I've had good success with the `gamma` distribution and the `suggest_initial()`-patch on my big data sets! I'm running it separately for ~200 experiments of 24 hours. Takes about 2h per experiment to both fit and estimate states with `viterbi()`, so it'll be running for a few days, but I'm stoked it's working well!

**Reply by TheoMichelot** [reply]
Okay that sounds good, thanks for the update. Out of curiosity, do you know roughly how long the model fitting and the Viterbi algorithm take, respectively? I would be interested to know how slow `viterbi()` is for large data sets (it is implemented in R, rather than C++ like model fitting).

**Reply by roaldarbol** [reply]
Absolutely! I was a bit curious, so I got some nicely formatted CLI output for it. 😉 So about 5 mins for the fitting, and 15 mins for the `viterbi()`, so it does take the majority or time spent.
```r
── Processing experiment from 2022-02-07 ───────────────────────────────────────
here() starts at /Volumes/ResearchSSD/002-Tracking-1Day
ℹ 2024-04-30 11:38:10 - Starting HMM for number: 1/6
ℹ 2024-04-30 11:38:40 - Fitting HMM
ℹ 2024-04-30 11:44:02 - Generating most likely sequence of states
ℹ 2024-04-30 11:59:16 - Starting HMM for number: 2/6
ℹ 2024-04-30 11:59:49 - Fitting HMM
ℹ 2024-04-30 12:06:12 - Generating most likely sequence of states
ℹ 2024-04-30 12:21:31 - Starting HMM for number: 3/6
ℹ 2024-04-30 12:22:04 - Fitting HMM
ℹ 2024-04-30 12:28:13 - Generating most likely sequence of states
ℹ 2024-04-30 12:42:55 - Starting HMM for number: 4/6
ℹ 2024-04-30 12:43:29 - Fitting HMM
ℹ 2024-04-30 12:49:47 - Generating most likely sequence of states
ℹ 2024-04-30 13:04:28 - Starting HMM for number: 5/6
ℹ 2024-04-30 13:05:05 - Fitting HMM
ℹ 2024-04-30 13:10:39 - Generating most likely sequence of states
ℹ 2024-04-30 13:25:21 - Starting HMM for number: 6/6
ℹ 2024-04-30 13:25:54 - Fitting HMM
ℹ 2024-04-30 13:31:32 - Generating most likely sequence of states
```

**Reply by TheoMichelot** [reply]
Very good to know, thanks Mikkel!

**Comment by roaldarbol** [comment]
Hi @TheoMichelot! Just came across this poster/paper from 2007 implementing a faster Viterbi algorithm, and it contains C++ code too. Maybe it's not too much hassle to implement at some point? Might at least make it easier. 😉 https://cs.idc.ac.il/~smozes/hmmspeedup/hmmspeedup.html

**Reply by roaldarbol** [reply]
Or here https://github.com/xiaoming-qxm/viterbi

**Reply by TheoMichelot** [reply]
Hi Mikkel, thanks for the references. Unfortunately, the challenge isn't to implement the Viterbi algorithm in C++ per se, but to combine TMB code with other C++ code in the package (see for example https://github.com/kaskr/adcomp/issues/247).

**Reply by roaldarbol** [reply]
Ah, that's why - thanks for the link too! Why does packaging always have to be hard... 😅
